ts0601_energy_meter quirk for Tuya power meter devices (v1 quirk)

jeverley opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->

Adds support for Tuya Energy Meter devices in 1 and 2 clamp configurations.

Uses zigbee2Mqtt https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/tuya.ts as reference for device datapoints (with amendments from community testing in the respective device support issues).

Meter devices providing power direction will report signed values for their supported attributes in accordance with the Zigbee cluster spec (instantaneous demand, active power and reactive power specifically).

**Provides control over metering sensor configuration:**

- The quirk allows the user to control the endpoint 3 power calculation on 2 clamp devices using the "channel_configuration" attribute value (calculated as either _a+b_ or _a-b_).
   - _a+b_ should be used when measuring total consumption from two power sources.
   - _a-b_ should be used to derive a second power value from input power and consumption i.e. clamp A on grid input, clamp B measuring power into the consumer unit.

Some devices report an a+b total natively, however in testing this value is often incorrect due to delay in detecting power direction change on the device.

In the real world when measuring both grid and production an a + b < 0 value would be atypical. 
This is mitigated using the _power_flow_preempt_ filter on device readings, this is optionally is enabled by setting 'power_flow_preempt' to True.


The quirk aims to make adding new devices simpler by utilising device specific dp_to_attribute mappings which feed into common core reporting clusters.

<img width="225" alt="image" src="https://github.com/zigpy/zha-device-handlers/assets/46714706/965aa80e-31ee-4105-bf78-b85e90a91664">

**Addresses device support requests:**

2 bidirectional clamps:

- https://github.com/zigpy/zha-device-handlers/issues/2549 (_TZE204_81yrt3lo)
- https://github.com/zigpy/zha-device-handlers/issues/2650 (_TZE200_rks0sgb7)

1 bidirectional clamp:
- https://github.com/zigpy/zha-device-handlers/issues/2420 (_TZE204_ac0fhfiq)

1 clamp:
- https://github.com/zigpy/zha-device-handlers/issues/1973 (_TZE204_cjbofhxw)

@jmuf I've just spotted that in parallel you've also been working on https://github.com/zigpy/zha-device-handlers/pull/2870 for  _TZE200_rks0sgb7, shall we pool ideas?

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

Associated with merged pre-requisite ZHA metering and electrical measurement changes:

- https://github.com/home-assistant/core/pull/107641
- https://github.com/home-assistant/core/pull/107576
- https://github.com/home-assistant/core/pull/109126

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [ ] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

jeverley committed:
ts0601_energy_meter quirk for Tuya power meter devices

Aims to address device support requests:

2 bidirectional clamps:
https://github.com/zigpy/zha-device-handlers/issues/2549 (_TZE204_81yrt3lo)
https://github.com/zigpy/zha-device-handlers/issues/2650 (_TZE200_rks0sgb7)

1 bidirectional clamp:
https://github.com/zigpy/zha-device-handlers/issues/2420 (_TZE204_ac0fhfiq)

1 clamp:
https://github.com/zigpy/zha-device-handlers/issues/1973 (_TZE204_cjbofhxw)

@jmuf I've just spotted that in parallel you've also been working on https://github.com/zigpy/zha-device-handlers/pull/2870 for 
_TZE200_rks0sgb7, shall we pool ideas?

jeverley committed:
Only some of the devices support the ZGP endpoint.

jeverley committed:
Linting corrections and updates to _TZE200_rks0sgb7

Amendments from details in https://github.com/zigpy/zha-device-handlers/issues/2650 & https://github.com/zigpy/zha-device-handlers/pull/2870 (however that approach uses attributes directly rather than datapoints).

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

jeverley committed:
Fix remaining lint errors

jeverley committed:
Update EARU datapoints based on @KPfff and @jmuf's findings

jeverley committed:
Move more functions into re-usable classes

jeverley committed:
Meter configuration defaults derived from device reporting channels

jeverley committed:
Use constants for reporting cluster attribute IDs, power reported with 1 decimal place

jmuf commented:
> @jmuf I've just spotted that in parallel you've also been working on https://github.com/zigpy/zha-device-handlers/pull/2870 for _TZE200_rks0sgb7, shall we pool ideas?

sure, tell  me what I should do. Your "quirk" looks more complete, my "quirk" is a one-off (and "my first quirk".. ).
* I had to resort to several "replacement clusters" to get the 3x same set of metrics (and buses to route data from the TUYA cluster to go there), that "feels wrong"
* also have still an issue with naming (all 3 measurements have same user-friendly name, didn't find way to label as Clamp1/Clamp2/Total)

Things I like in "mine":
* using the constant from ElectricalMeasurement, Metering (instead of random HEX values in the quirk)
* my quirk (or my meter?) seems to happily report negative values, so your complicated configuration (a+b / a-b) seems not required (also a user could just turn around the clamp to reverse the reading?)
* double-digit precision on "Watt" sounds not what these meters are made for. Mine says ±2W <100W, else ±2% accuracy

jeverley commented:
> sure, tell me what I should do. Your "quirk" looks more complete, my "quirk" is a one-off (and "my first quirk".. ).
> 
> * I had to resort to several "replacement clusters" to get the 3x same set of metrics (and buses to route data from the TUYA cluster to go there), that "feels wrong"
> * also have still an issue with naming (all 3 measurements have same user-friendly name, didn't find way to label as Clamp1/Clamp2/Total)
> 
> Things I like in "mine":
> 
> * using the constant from ElectricalMeasurement, Metering (instead of random HEX values in the quirk)
> * my quirk (or my meter?) seems to happily report negative values, so your complicated configuration (a+b / a-b) seems not required (also a user could just turn around the clamp to reverse the reading?)
> * double-digit precision on "Watt" sounds not what these meters are made for. Mine says ±2W <100W, else ±2% accuracy

Appreciate the feedback @jmuf , I've started to incorporate some changes based on your thoughts.

Changes so far:

- Switched to contestants for the attribute IDs
- Amend power divisor to report 1 decimal place (one of the meters reports that level of precision)

Outstanding:

- _also have still an issue with naming (all 3 measurements have same user-friendly name, didn't find way to label as Clamp1/Clamp2/Total)_
I like the idea, looking at the core integration code I think we need to make some amendments to ZHA's sensor functionality to support this.
- _my quirk (or my meter?) seems to happily report negative values, so your complicated configuration (a+b / a-b) seems not required (also a user could just turn around the clamp to reverse the reading?)_
  - Ah that's useful to know yours reports the signed integers natively. 
  The other devices report power as an unsigned integer, so we need to apply the direction locally. I can make that logic only apply for the uint devices.
  - On the add/subtract point, that's a fair challenge, it does add additional complexity.
  I still think there's benefit in a-b for non-bidirectional 2 clamp devices because that would be the only way of deriving export to the grid (i.e. consumption minus solar, since both should clamps should always be positive).

Xesc1714 commented:
Hi, 
Some ZHA quirks to setup and work with this device TZ3000_zrm3oxsh TS011F
<img width="397" alt="image" src="https://github.com/zigpy/zha-device-handlers/assets/118026358/702d7262-9757-42c6-8031-e6670f1fd5bf">
thank you

pantonvich commented:
Hey thanks for this quirk! I have the _TZE204_81yrt3lo but with only one clamp - some how I managed to order it with only one clamp - lol. My question is that there doesn't appear to be a difference between the power and instantaneous demand so why are both needed? I set mine up on the A and did have to flip my clamp to not get negative watts.

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

jeverley committed:
Refactor following tests over the last few weeks

jeverley committed:
Fix compatibility with python <3.10

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

jeverley committed:
MeasurementType defined in quirk until zcl cluster class has correct values.

jeverley committed:
Power flow preempt should occur before directional attribute handling

jeverley committed:
Add typing hints to functions and refactor virtual and preempt classes

jeverley committed:
Initial a_minus_b summation attr virtual channel implementation

jeverley committed:
Correct storage of previous value

jeverley committed:
Improvements to virtual channel value calculations

Testing showed some assumptions to be incorrect, consumption_minus_production still needs some work

jeverley committed:
Forcing an empty commit.

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

jeverley committed:
Improve config defaults handling

jeverley committed:
Code structure improvements

jeverley committed:
Additional refactor in support of consumption_minus_production summation reporting

Vladdy65 commented:
Following. Just acquired one of the dual clamp  _TZE204_81yrt3lo units and would dearly like to get it working. Happy to help with any testing.

jeverley commented:
> Following. Just acquired one of the dual clamp  _TZE204_81yrt3lo units and would dearly like to get it working. Happy to help with any testing.

Your help would be much appreciated @Vladdy65.
I think we're nearly there functionality wise if you wanted to give the current version in the PR a go on your end.

Current observations indicate that you probably won't see all sensors show up until after a restart of the zigbee integration due to how home assistant currently determines if a sensor is available - something I plan to tackle.

I'm also still making some final corrections to how the sensor calculates the grid summation values in the consumption minus production configuration, hoping this coming week I have time to address these remaining pieces.

Vladdy65 commented:
@jeverley sure thing.

I used the following quirks file: [https://github.com/jeverley/zha-device-handlers/blob/ts0601_energy_meter/zhaquirks/tuya/ts0601_energy_meter.py](url) downloaded in the wee hours of this morning.
I use the SkyConnect v1.0 zigbee dongle.

Setup procedure:

- Ensured my configuration.yaml included the following:

`zha:
  database_path: /config/zigbee.db
  enable_quirks: true
  custom_quirks_path: /config/custom_zha_quirks/`

- Removed device from HA Zigbee Home Automation (ZHA)
- Deleted all quirks files and _pycache_ folder from config/custom_zha_quirks/ folder
- Restarted HA
- Dropped the ts0601_energy_meter.py file in the config/custom_zha_quirks/ folder
- Restarted HA
- Added the Dual Channel Smart Meter to ZHA (ADD DEVICE, hold button on unit for 5s)

Got no Sensors only the Configuration and Diagnostic panels.

- Rebooted HA

Success!! 
Got the below (having also enabled the RSSI Diagnostic sensor): 
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/e0854589-25a1-47c6-9ac4-67d076cba8e1)

At this point it was under my desk with the clamps not connected to anything.

- Mounted the unit in the distribution cabinet and connected the clamps, following the arrow directions inside pointing from supply side to each of my two consumer cabinets incoming Live wires.

After a moment I got the below:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/d3f5638a-68cc-4817-8533-94089138c5f0)

**Fantastic work so far @jeverley !**

A few observations.

- Sign for power 'received' from mains supply in my view is correct and should be negative (-) Power, it is power that has been consumed vs. power that has been generated, say from solar which logically should be a positive number. Suspect some will not fancy plotting their consumption as a -ve number and I'm not sure how this plays with the HA Power components. Not had time to test.
- The Summation delivered of 0.06kWh is probably due to me having one of the clamps the wrong way round for a short period of time.
- The _Unknown_ AC frequency and Voltage are probably surplus as from what I understand the device only measures this off its own supply feed, not the clamps.
- The same probably applies to the _Unknown_ Current as this is only measured off the two clamps, unless you have plans to make this the sum of the A+B channels.
- The third _Unknown_ Power factor, again, seems redundant. Power factor is measured as the shift in the phase of the current vector (as phase of the voltage signal from each clamp) vs the phase of the voltage signal on the unit's power supply. So there can only be two measurements. Unless I'm missing something.

My ZHA diagnostic file attached:
[config_entry-zha-1602f4cec172a9c30a09aa4ae0e5f2a0 (1).json](https://github.com/zigpy/zha-device-handlers/files/14807368/config_entry-zha-1602f4cec172a9c30a09aa4ae0e5f2a0.1.json)

Let me know if yo need anything else tested.

jeverley commented:
Thanks very much @Vladdy65, please could you give the version in branch https://github.com/jeverley/zha-device-handlers/tree/ts0601_energy_meter_devices a try? 

File linked directly:
https://github.com/zigpy/zha-device-handlers/blob/ts0601_energy_meter_devices/zhaquirks%2Ftuya%2Fts0601_energy_meter.py

That one has my more recent changes including the different configurable summation methods and support for similar device models.

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

reef-actor commented:
> power 'received' from mains supply

The terminology is refering to the load (aka your house). When you are using electricity it is being "delivered" to your house, when you are producing it is being "received" from your house.

Vladdy65 commented:
@jeverley
> File linked directly:
> https://github.com/zigpy/zha-device-handlers/blob/ts0601_energy_meter_devices/zhaquirks%2Ftuya%2Fts0601_energy_meter.py

The link gives me a 404 error.

Could you confirm again correct file path please.

jeverley commented:
Sorry about that, here you go:

https://raw.githubusercontent.com/jeverley/zha-device-handlers/ts0601_energy_meter_devices/zhaquirks/tuya/ts0601_energy_meter.py

Vladdy65 commented:
> The terminology is refering to the load (aka your house). When you are using electricity it is being "delivered" to your house, when you are producing it is being "received" from your house.

@reef-actor 
The way I interpreted it is, using 'House' as the subject in the sentence:
- House 'received' power from grid;
- House 'delivered' power to grid - e.g. when you have solar and you are selling it back to the grid;

I tried but struggled to make sentences using 'House' as the subject with the terms the other way round.
I came up with:
- House has power 'delivered' to it from the grid;

So I asked our friendly AI (Copilot) about it, this is the conversation 🤔:

**You**
Please explain the semantics of this sentence with respect of the nouns and verbs and who is the subject: House has power 'delivered' to it from the grid
**Copilot**
Certainly! Let’s break down the semantics of the sentence:

Subject:
The subject of the sentence is “House.”
The subject is the entity that the sentence is primarily about.
Verbs:
The main verb in the sentence is “has delivered.”
The auxiliary verb is “has.”
The verb phrase “has delivered” indicates an action or state related to the subject.
Nouns:
The nouns in the sentence are:
“House”: Refers to a building or dwelling where people live.
“power”: Refers to electrical energy.
“grid”: Refers to the electrical power distribution network.
Semantics:
The sentence conveys that the “House” **receives** or obtains “power” from the “grid.”
The action of delivering power is attributed to the grid.
In other words, the grid supplies electricity to the house.
In summary, the subject (house) benefits from the action of power delivery facilitated by the grid. 🏠⚡️

reef-actor commented:
A better reference might be the [zigbee standard (Page 217)](https://zigbeealliance.org/wp-content/uploads/2019/11/docs-07-5356-19-0zse-zigbee-smart-energy-profile-specification.pdf) 😜

Vladdy65 commented:
> A better reference might be the [zigbee standard (Page 217)](https://zigbeealliance.org/wp-content/uploads/2019/11/docs-07-5356-19-0zse-zigbee-smart-energy-profile-specification.pdf) 😜

@reef-actor I stand humbled by my ignorance. 😔

Vladdy65 commented:
@jeverley 
With:
> https://raw.githubusercontent.com/jeverley/zha-device-handlers/ts0601_energy_meter_devices/zhaquirks/tuya/ts0601_energy_meter.py

No change on first adding the device:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/200210a0-d59a-4ac0-a74a-77b967aa29d2)

After restarting HA:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/6dbfc3fe-eae5-47d0-8415-ccc265d0a9e6)

Seems like some weird stuff is going on now I'm afraid.
At first I thought HA was doing something odd, so re-did the whole exercise and made sure I renamed the device (I'd given it the same name as before initially), but same outcome.
Let me know when you need further help.

jeverley commented:
If you go into Manage Device, please could you read the values of the attributes:

- channel_configuration
- power_flow
- power_flow_ch_b
- power_flow_preempt
- suppress_reverse_flow
- suppress_reverse_flow_ch_b

From the TuyaEnergyMeterManufCluster?

Vladdy65 commented:
What I have is:

- Clusters
TuyaEnergyMeterManufCluster_2CHB_MatSeePlus (Endpoint id: 1, Id: 0xef00, Type: in)

- channel_configuration (id: 0x5000)
ChannelConfiguration_2CHB.GRID_PLUS_PRODUCTION
- power_flow (id: 0x0266)
PowerFlow.REVERSE
- power_flow_ch_b (id: 0x0268)
PowerFlow.REVERSE
- power_flow_preempt (id: 0x5020)
Bool.true
- suppress_reverse_flow (id: 0x5010)
None
- suppress_reverse_flow_ch_b (id: 0x5011)
None

jeverley commented:
> What I have is:
> 
> - Clusters
> TuyaEnergyMeterManufCluster_2CHB_MatSeePlus (Endpoint id: 1, Id: 0xef00, Type: in)
> 
> - channel_configuration (id: 0x5000)
> ChannelConfiguration_2CHB.GRID_PLUS_PRODUCTION
> - power_flow (id: 0x0266)
> PowerFlow.REVERSE
> - power_flow_ch_b (id: 0x0268)
> PowerFlow.REVERSE
> - power_flow_preempt (id: 0x5020)
> Bool.true
> - suppress_reverse_flow (id: 0x5010)
> None
> - suppress_reverse_flow_ch_b (id: 0x5011)
> None

Thanks @Vladdy65, do you definitely have the arrow inside the clamps facing in the direction of delivered power, and the red cables of the clamps connected to their corresponding S1 terminal on the module?

Vladdy65 commented:
> Thanks @Vladdy65, do you definitely have the arrow inside the clamps facing in the direction of delivered power, and the red cables of the clamps connected to their corresponding S1 terminal on the module?

@jeverley double checked the wiring. Also made sure I had the LIVE cables (brown in UK). 
To confirm, I followed this guidance:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/7a0a173d-9df5-4945-a8f6-5fd5429cafa6)
In my case the arrow is inside the clamp, K->L, as in the image inset, and is pointing to the loads as follows:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/9b832c03-44a6-48b3-a49c-1ef5c48fbbfd)

**Wiring:**
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/77af78d3-3a01-441b-a01e-6297da271d33)

jeverley commented:
Appreciate the thoroughness, if you flip the orientation of the clamps does the flow direction change?

reef-actor commented:
I have a different meter model with a single clamp (CT-2101A) that is labelled as "S1.Black, S2.Red", does your manual say the other way round? I have tried looking up how these clamps work and I am not sure they would do anything at all if they were the wrong way round, but I thought I would mention it.

Vladdy65 commented:
> I have a different meter model with a single clamp (CT-2101A) that is labelled as "S1.Black, S2.Red", does your manual say the other way round? I have tried looking up how these clamps work and I am not sure they would do anything at all if they were the wrong way round, but I thought I would mention it.

No mention of it in my manual.

Vladdy65 commented:
> Appreciate the thoroughness, if you flip the orientation of the clamps does the flow direction change?

Before, with clamps pointing to Load:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/d76f9b89-16ab-4e24-af6a-55eb323c8ad0)

After, with clamps in opposite direction:
![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/6b9ab854-8103-43de-bd50-5c9041b944e7)

Only significant change I see is in the Power Factor. I had to ponder this and brush up on my power factor equations, but I am no wiser as to what just happened. Other than a PF of 39% is not good. I could accept it on channel B where the current is only 0.2A and thus probably close to measurement limits (it is an 80A clamp), but that it flipped and now shows it on channel A is surprising.

jeverley commented:
Hmm that's odd, I'd have expected a flow reversal in that scenario - are you able to read the app_version attribute value from the Basic cluster?
That should show if we're running the same firmware version, I've got 74 (or 0x4a).

I can add some debug lines in the quirk to write out the raw reported data point values, possible your device is behaving differently to my own.

Vladdy65 commented:
This morning the picture looked different, and more in line with expectations of signs changing. I've left the clamps on 'backward' pointing away from Load. I'm now wondering if it needed some time to 'settle' after changing the clamp direction before taking the 'After' screenshot last night. 
As for app_version attribute value from the Basic cluster, I get no response.

![image](https://github.com/zigpy/zha-device-handlers/assets/159207250/9a730583-938c-40cf-83bf-dea67e5df225)
and it has been taking readings through the night as I had a history graph going to confirm. But now it is frozen. I've removed it but cannot add it back again....

Is there any way to factory reset these devices?

Vladdy65 commented:
Its back...
Basic (Endpoint id: 1, Id: 0x0000, Type: in)
app_version (id: 0x0001)
74
Same as yours.

decrouxtom commented:
Thanks for the answer @reef-actor for https://github.com/zigpy/zha-device-handlers/issues/2420#issuecomment-2028916235
Power seems ok:
![Screenshot_2024-04-01-11-27-13-783_io homeassistant companion android-edit](https://github.com/zigpy/zha-device-handlers/assets/84731174/2f6818b4-470f-4306-b50b-cfeaeaf88141)

With the new quirk, total energy seems wrong:
![Screenshot_2024-04-01-11-52-38-862_io homeassistant companion android-edit](https://github.com/zigpy/zha-device-handlers/assets/84731174/7e5a2964-210a-4c65-b841-e88eab69b73a)

andrewoe commented:
Hi folks,
I really struggling with the same device....
Did all as described here (e.g. tried several quirks from this side), but I got no sensors, anyway...
I have several Zigbee devices in place, such as of thermostates.
What I wondering is the fact, that there is no pycache file in the zha_quirks folder...
Any ideas, what I missed or may misunderstood?
thanks in advance
Andy

ml0renz0 commented:
Hi guys,

2-clamp device here, same scenario as @Vladdy65 

Clamps are set in the proper direction but they are reading negative values just as @Vladdy65 experienced.

Should I reverse the direction of the clamp?

PS: thanks for the great job @jeverley!

Dominic-070 commented:
> Sorry about that, here you go:
> 
> https://raw.githubusercontent.com/jeverley/zha-device-handlers/ts0601_energy_meter_devices/zhaquirks/tuya/ts0601_energy_meter.py

Where do i add this?

Dominic-070 commented:
> Sorry about that, here you go:
> 
> https://raw.githubusercontent.com/jeverley/zha-device-handlers/ts0601_energy_meter_devices/zhaquirks/tuya/ts0601_energy_meter.py

I have applied this.
it appears that the amperages and VA of both CT terminals are correct, but the wattages are not. The values ​​surrounded in green seem to be correct, the rest are not.
I am trying to measure the consumption of 2 individual devices. So no return only consumption,

![image](https://github.com/user-attachments/assets/0c2c3ad6-c443-4eb3-9877-5cd771c87ec6)

drug123 commented:
> it appears that the amperages and VA of both CT terminals are correct, but the wattages are not. 
Same for me -  there is no way I have consumed 3 megawatts in two months :))

Dominic-070 commented:
> > it appears that the amperages and VA of both CT terminals are correct, but the wattages are not.
> > Same for me -  there is no way I have consumed 3 megawatts in two months :))

I'm afraid this quirk is specially made for PV distributors with consumption on one side and production on the other. Unfortunately, I have not yet been able to find a good quirk to measure the consumption of the 2 individual devices with the TS0601 _TZE204_81yrt3lo

Rivenage commented:
I've been successful in getting this quirk loaded showing all the sensor attributes, but am not getting any data except for frequency and voltage. I've related ZHA, deleted and readded the zigbee device, rebooted the HA, verified wiring, tried flipping the clamp and wires, nothing seems to work. Any suggestions?
![Screenshot_20240730-222822](https://github.com/user-attachments/assets/8d1e231d-a4cb-4dfb-9d88-f3b9cf4a3a53)

Rivenage commented:
I had the same issue with positive and negative values. With the clamps mounted in the correct orientation based on the arrows by reversing the wires at the sensor box, I was able to get all the values populating correctly.
![Screenshot_20240731-070034](https://github.com/user-attachments/assets/85697ab4-18ce-4a1b-a186-117da8569075)
![PXL_20240731_110025286](https://github.com/user-attachments/assets/cc0cdefc-4af6-422f-abbe-cbfc94cdb98c)

jeverley committed:
Merge branch 'zigpy:dev' into ts0601_energy_meter_devices

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

michaelschnyder commented:
I'm working on adding _TZE284_81yrt3lo and plan to base my work on branch from you @jeverley. 

Zigbee2Mqtt folks seem to got it working based on 
* the discussion on https://github.com/zigpy/zha-device-handlers/issues/3658
* and notes on https://github.com/KiwiHC16/Abeille/blob/5db0421a7afbb83d130b8832fd48bafbfc807dc4/core/config/devices/TS0601__TZE204_81yrt3lo/notes.txt#L4

Di you plan to get this one merged upstream anytime soon, or would it make more sense for me to base it on the QuickBuilder?

**My Setup**
* Single clamp available: MatSeePlus PJ-1203A (on the case) vs. PJ-1203AN (in the box)
* Local setup with HAOS in Proxmox

jeverley commented:
> I'm working on adding _TZE284_81yrt3lo and plan to base my work on branch from you @jeverley. 
> 
> Zigbee2Mqtt folks seem to got it working based on 
> * the discussion on https://github.com/zigpy/zha-device-handlers/issues/3658
> * and notes on https://github.com/KiwiHC16/Abeille/blob/5db0421a7afbb83d130b8832fd48bafbfc807dc4/core/config/devices/TS0601__TZE204_81yrt3lo/notes.txt#L4
> 
> Di you plan to get this one merged upstream anytime soon, or would it make more sense for me to base it on the QuickBuilder?
> 
> **My Setup**
> * Single clamp available: MatSeePlus PJ-1203A (on the case) vs. PJ-1203AN (in the box)
> * Local setup with HAOS in Proxmox

I'd started to lay the groundwork for a quirk builder based implementation over the weekend using the tuya quirk builder component (https://github.com/zigpy/zha-device-handlers/blob/dev/tuya.md), I'll push these local changes to a new branch this afternoon.

I think we can simplify a lot of what's in my existing v1 quirk, particularly to get a first release of it mainlined.

jeverley commented:
Work in progress on the v2 quirk can be found here, should have something testable locally by tomorrow - note this will definitely produce errors right now.

https://github.com/jeverley/zha-device-handlers/blob/ts0601_energy_meter_v2_quirk/zhaquirks/tuya/ts0601_energy_meter.py

michaelschnyder commented:
Great @jeverley! I only was able to get channel A to work, with no modifications, just adding the device information. 
Was not able to get Channel B to work, which also deos not not work in Z2M (my current setup)

jeverley commented:
> Great @jeverley! I only was able to get channel A to work, with no modifications, just adding the device information. 
> Was not able to get Channel B to work, which also deos not not work in Z2M (my current setup)

The code in the [v2 repo](https://github.com/jeverley/zha-device-handlers/blob/ts0601_energy_meter_v2_quirk/zhaquirks/tuya/ts0601_energy_meter.py) should now expose the device using the new quirk format. I currently can see values from both clamps in my testing, and you can select different virtual channel calculations (A plus B, A minus B, B minus A).

There's still some work needed on the power flow handling, but you're welcome to give it a try.
Note you'll likely see quite a few unavailable attributes if you don't first remove and re-pair the device prior to using the V2 quirk.

lh-128 commented:
I have a working handler for TZE284 based on @jeverley 's work.

https://github.com/lh-128/home-assistant/blob/main/Tuya_TZE284_81yrt3lo.py

While waiting for his refactoring to complete, you can use mine in the interim. It should work for both channels. Tested working for me.

> I'm working on adding _TZE284_81yrt3lo and plan to base my work on branch from you @jeverley.
> 
> Zigbee2Mqtt folks seem to got it working based on
> 
> * the discussion on [[Device Support Request] TS0601 _TZE284_81yrt3lo Tuya Smart Life Dual channel energy meter #3658](https://github.com/zigpy/zha-device-handlers/issues/3658)
> * and notes on https://github.com/KiwiHC16/Abeille/blob/5db0421a7afbb83d130b8832fd48bafbfc807dc4/core/config/devices/TS0601__TZE204_81yrt3lo/notes.txt#L4
> 
> Di you plan to get this one merged upstream anytime soon, or would it make more sense for me to base it on the QuickBuilder?
> 
> **My Setup**
> 
> * Single clamp available: MatSeePlus PJ-1203A (on the case) vs. PJ-1203AN (in the box)
> * Local setup with HAOS in Proxmox

jeverley commented:
I'll be progressing this under the new v2 quirk PR https://github.com/zigpy/zha-device-handlers/pull/3824
